File: leetcode/0098-validate-binary-search-tree/solution.py
# Definition for a binary tree node.
# class TreeNode:
#     def __init__(self, val=0, left=None, right=None):
#         self.val = val
#         self.left = left
#         self.right = right
class Solution:
    def isValidBST(self, root: TreeNode) -> bool:
        return self.isValidBSTUtil(root, None, None)
    
    def isValidBSTUtil(self, root: TreeNode, minVal: int, maxVal: int) -> bool:
        if root == None:
            return True
        
        if minVal != None and root.val <= minVal:
            return False
        if maxVal != None and maxVal <= root.val:
            return False
        
        return self.isValidBSTUtil(root.left, minVal, root.val) and self.isValidBSTUtil(root.right, root.val, maxVal)

File: leetcode/0098-validate-binary-search-tree/solution.cpp
/**
 * Definition for a binary tree node.
 * struct TreeNode {
 *     int val;
 *     TreeNode *left;
 *     TreeNode *right;
 *     TreeNode(int x) : val(x), left(NULL), right(NULL) {}
 * };
 */
class Solution {
public:
    bool isValidBST(TreeNode* root, int *minVal, int *maxVal) {
        if (root == nullptr)
            return true;
        if (minVal != nullptr && root->val <= *minVal)
            return false;
        if (maxVal != nullptr && *maxVal <= root->val)
            return false;
        
        return isValidBST(root->left, minVal, &root->val) && isValidBST(root->right, &root->val, maxVal);
    }
    
    bool isValidBST(TreeNode* root) {
        return isValidBST(root, nullptr, nullptr);
    }
};